[16.0] purchase_packaging_default: default packaging return first pac…

cyrilmanuel opened the pull request:
…kaging with purchase=True


This PR allow to return the first packaging on the list of packaging_ids, that meet the condition purchase = true

santostelmo reviewed (approved):
LG. 
It would be a plus if you write a quick unit test: Test that only purchasable packaging are added to PO line

cyrilmanuel commented:
Hello, 

today the test contain the right order to trigger the carton as first :)

i can add a test with a wrong example to check if its not working correctly. what do you think ? @santostelmo

lmignon reviewed (approved):
Great. Can you add a test to ensure that this feature will not be lost in the future.

lmignon commented on purchase_packaging_default/models/purchase_order_line.py:
```diff
@@ -12,7 +12,9 @@ class PurchaseOrderLine(models.Model):
         """From product get 1st packaging found ordered by sequence"""
         product_template = self.product_id.product_tmpl_id
-        return fields.first(product_template.packaging_ids)
+        return fields.first(
+            product_template.packaging_ids.filtered(lambda p: p.purchase)
+        )
```
This approach is not optimal... You first loop over all the packaging and then return the first one.... 
```suggestion
        for pkg in  product_template.packaging_ids:
            if pkg.purchase:
                return pkg
        return product_template.packaging_ids.browse()
```

HviorForgeFlow reviewed (changes requested):
@cyrilmanuel can you attend comment, thanks: Attend https://github.com/OCA/purchase-workflow/pull/2315#discussion_r1662613500

santostelmo commented:
@cyrilmanuel could you please have a look on this one ?

santostelmo committed:
purchase_packaging_default: default packaging return first packaging with purchase=True

santostelmo committed:
purchase-packaging-default-packaging: computed default product packaging

We assign default product packaging via api.depends as we need to compute product packaging even if PO is not created from form view. For example reordering rules create Pos and onchange is not triggered

santostelmo commented:
@HviorForgeFlow I reworked this PR:
1. Add [change suggested](https://github.com/OCA/purchase-workflow/pull/2315#discussion_r1662613500) by @lmignon 
2. Add a new commit: purchase-packaging-default-packaging: computed default product packaging. We assign default product packaging via api.depends as we need to compute product packaging even if PO is not created from form view. For example reordering rules create PO's and onchange is not triggered

Could you please review again ?

lmignon reviewed (changes requested):
Thank you for the PR but IMO this kind of changes deserve some unit tests. 
* Only purchase packaging are used and set on onchange 
* If a packing is specified into the data given to the create method, it's used otherwise if a default one exists this last one is used.
* ???

santostelmo commented:
> Thank you for the PR but IMO this kind of changes deserve some unit tests.
> 
> * Only purchase packaging are used and set on onchange
> * If a packing is specified into the data given to the create method, it's used otherwise if a default one exists this last one is used.
> * ???

Preparing some unit tests

santostelmo commented:
@lmignon added a unit test to cover the following:
1. setting is enabling desired behavior => get the first product packaging in the sequence 
2. PO created from directly from api still have default packaging computed

Let me know if you still see any other case that should be covered

Ricardoalso commented on purchase_packaging_default/tests/test_purchase_packaging_default.py:
```diff
@@ -39,3 +41,56 @@ def test_purchase_packaging_default(self):
             line_f.product_qty = 241
             self.assertEqual(line_f.product_packaging_qty, 2)
+
+    # def _find_suitable_product_packaging(self, product_qty, uom_id):
+    #         """ try find in `self` if a packaging's qty in given uom is a divisor of
+    #         the given product_qty. If so, return the one with greatest divisor.
+    #         """
+    #         packagings = self.sorted(lambda p: p.qty, reverse=True)
+    #         for packaging in packagings:
+    #             new_qty = packaging._check_qty(product_qty, uom_id)
+    #             if new_qty == product_qty:
+    #                 return packaging
+    #         return self.env['product.packaging']
```
```suggestion

```

santostelmo replied:
Thank you @Ricardoalso . Fixed

Ricardoalso reviewed (approved):
LGTM

lmignon reviewed (approved):
LGTM (Code + Functionnal)

ivantodorovich reviewed (approved):
LG

santostelmo commented:
@HviorForgeFlow good for you ?

santostelmo commented:
> @HviorForgeFlow good for you ?

@HviorForgeFlow Could you kindly review the change request?

HviorForgeFlow commented:
/ocabot merge minor
